tests/h/services/search_index/_queue_test.py

```diff
@@ -351,35 +336,96 @@ def search_index(
-                search_index.add_annotation(annotation)
+        By default this creates jobs with job.name="sync_annotation", with a
+        scheduled_at time in the past, and with a job.kwargs that contains an
+        annotation_id and force=False.
 
-            es_client.conn.indices.refresh(index=es_client.index)
+        By default a new annotation will be created for the job to use.
 
-        return index
+        By default the annotation will exist in the DB but will *not* be in
+        Elasticsearch.
 
-    @pytest.fixture
-    def add_all(self, queue, annotation_ids):
-        def add_all(ids=annotation_ids, schedule_in=MINUS_5_MIN_IN_SECS, force=False):
-            for id_ in ids:
-                queue.add_by_id(
-                    id_, tag="test_tag", schedule_in=schedule_in, force=force
-                )
+        Usage:
+
+            SyncAnnotationJobFactory()
+
+            # Create multiple jobs, each with its own annotation.
+            SyncAnnotationJobFactory.create_batch(size=3)
+
+            # Create a job that isn't scheduled yet.
+            one_hour_from_now = datetime.datetime.utcnow() + datetime.timedelta(hours=1)
+            SyncAnnotationJobFactory(scheduled_at=one_hour_from_now)
+
+            # Create an expired job.
+            one_hour_ago = datetime.datetime.utcnow() - datetime.timedelta(hours=1)
+            SyncAnnotationJobFactory(expires_at=one_hour_ago)
+
+            # Create multiple jobs with force=True in their kwargs.
+            SyncAnnotationJobFactory.create_batch(size=3, force=True)
+
+            # Create a job for a given annotation, instead of
+            # SyncAnnotationJobFactory creating a new annotation automatically.
+            annotation = factories.Annotation()
+            SyncAnnotationJobFactory(annotation=annotation)
+
+            # Create a job for an annotation that *is* already in Elasticsearch.
+            SyncAnnotationJobFactory(index=True)
+        """
+
+        class SyncAnnotationJobFactory(factories.Job):
+            class Meta:
+                exclude = "force"
+
+            annotation = factory.SubFactory(factories.Annotation)
+            force = False
+
+            scheduled_at = factory.LazyFunction(
+                lambda: now - datetime_.timedelta(hours=1)
+            )
+            name = "sync_annotation"
+
+            kwargs = factory.LazyAttribute(
+                lambda o: {
+                    "annotation_id": URLSafeUUID.url_safe_to_hex(o.annotation.id),
+                    "force": o.force,
+                }
+            )
+
+            @classmethod
+            def _create(cls, model_class, *args, **kwargs):
+                annotation = kwargs.pop("annotation")
+                index_ = kwargs.pop("index", False)
+
+                job = super()._create(model_class, *args, **kwargs)
+
+                if index_:
+                    index(annotation)
+
+                return job
+
+        return SyncAnnotationJobFactory
```

Comment:
The downside of all this is that you need a reasonably clever factory: it has to use `exclude` and has to override `_create()`. It's pretty simple though and I think understanding factory_boy is the kind of knowledge that will pay dividends, since we use factory_boy everywhere.

The factory class also has to be defined inside a fixture because it depends on the `index` fixture :man_shrugging:

Reply:
It might honestly be better if some the cleverness were removed from this factory and it just required each test to create and index the annotation itself:

```python
def test_foo(self, factories, index, SyncAnnotationJobFactory):
    annotation = factories.Annotation()
    index(annotation)
    job = SyncAnnotationJobFactory(annotation)
    
    ...
```

That turns one line into three in every test (or 2 if the test doesn't need the annotation to be indexed). But it's arguably clearer because you can plainly see in the test that an annotation was added to the DB and indexed into Elasticsearch and then a job for that annotation was created. Rather than having to know that `SyncAnnotationJobFactory` does all that. And it'd remove all the cleverness from `SyncAnnotationJobFactory`.